### ISSUE TITLE
Remove tracer & context from Span

### ISSUE THREAD

**Issue by delner** [DataDog/dd-trace-rb#1514]
## The issue

`Datadog::Span` currently is aware of the `Tracer` and `Context`. Both of these components are upstream of spans (the tracer "creates spans & manages context" and the context "collects spans"); the Span itself is merely a data structure and should not reference these (doing so introduces circular references.)

## Analysis

`context` and `tracer` are likely a part of the `Span` because of the trace API:

```ruby
span = tracer.trace('my.operation')
do_work
span.finish
```

Because you can call `finish` on the span directly, it bypasses the tracer/context. Since those need to be closed properly, `Span` has to call them.

Within `Span`, `Tracer` and `Context` specifically appears in these two places:

The initializer:

```ruby
def initialize(tracer, name, options = {})
  @tracer = tracer
  @context = options.fetch(:context, nil)
  # ...
end
```

The `finish` method:

```ruby
# Mark the span finished at the current time and submit it.
def finish(finish_time = nil)
  # A span should not be finished twice. Note that this is not thread-safe,
  # finish is called from multiple threads, a given span might be finished
  # several times. Again, one should not do this, so this test is more a
  # fallback to avoid very bad things and protect you in most common cases.
  return if finished?

  @allocation_count_finish = now_allocations

  now = Utils::Time.now.utc

  # Provide a default start_time if unset.
  # Using `now` here causes duration to be 0; this is expected
  # behavior when start_time is unknown.
  start(finish_time || now) unless started?

  @end_time = finish_time || now
  @duration_end = finish_time.nil? ? duration_marker : nil

  # Finish does not really do anything if the span is not bound to a tracer and a context.
  return self if @tracer.nil? || @context.nil?

  # spans without a service would be dropped, so here we provide a default.
  # This should really never happen with integrations in contrib, as a default
  # service is always set. It's only for custom instrumentation.
  @service ||= (@tracer && @tracer.default_service)

  begin
    @context.close_span(self)
    @tracer.record(self)
  rescue StandardError => e
    Datadog.logger.debug("error recording finished trace: #{e}")
    Datadog.health_metrics.error_span_finish(1, tags: ["error:#{e.class.name}"])
  end
  self
end
```

## Suggestion

Remove the `tracer` and `context` references from the `Span` then have the `Tracer` do the work at the end of `finish` (e.g. update the context, record the span, etc)

Update the trace API such that any instrumentation must call `tracer.finish_span(span)` instead of `span.finish`. It's not as "pretty" but is a much simpler technical design as all the context management can then be done within `Tracer` (better isolation.)

**Comment by marcotc** [comment]
My concern with `tracer.finish_span(span)` is that you could finish a span on a different context then it was started.

```ruby
span = Datadog.tracer.trace('my.operation')
Datadog.configure {}
Datadog.tracer.finish_span(span)
```

the hard references that the span holds to the context and tracer today prevent that from happening.

**Comment by delner** [comment]
There's nothing preventing a span being finished in a different context today anyways. I think reconfiguring the tracer mid-trace is an outlier edge case in which it would be okay to drop the span/trace.

**Comment by marcotc** [comment]
`UnfinishedSpan` sounds reasonable.

My concern during reconfiguration in-between a span start/finish is not about precise instrumentation, but about the correct disposure of resources.

Even if the old thread-local context won't be flushed to the backend, I think we should still finish the spans it created on that original context. Otherwise we are left with lingering objects.

They might not be an issue, but I think the pattern of creating and disposing of resources correctly, even if they are not useful anymore to be good hygiene. I see it similar to having an open file descriptor in memory, for a file that has already been removed from the file system: there's no impact in the long term for not properly disposing of that file, but closing it properly ensures that we are correctly matching creation and disposition operations.

Despite that, if the cost of correctly disposing of obsolete objects is too high, I can see ourselves not adhering to a strict create/dispose pattern. I don't see the implementation of `UnfinishedSpan` to be a large burden, so at this point I'm happy with the `UnfinishedSpan` solution.

**Comment by lloeki** [comment]
I like the `UnfinishedSpan` move: intuitively I shared the same concern as @marcotc, as I felt it would make `Datadog.tracer` carry a bit too much of implicit global responsibilities, which `UnfinishedSpan` appears to encapsulate more explicitly.

> Doing it this way could even avoid the breaking API change.

Are there places where `foo.is_a?(Datadog::Span)` is expected to happen? I have a feeling that `class UnfinishedSpan` should `< Span` somehow, or at least pretend to be, as `def_delegators  :span, *Datadog::Span.instance_methods` seems to suggest.

**Comment by delner** [comment]
> Are there places where foo.is_a?(Datadog::Span) is expected to happen?

@lloeki This does happen in a few places; namely in `Tracer#record` and in several tests. I'm not sure its possible for `UnfinishedSpan < Span` because its meant to wrap a span with its context, not be a span itself. It's really more of an `UnfinishedTraceOperation` with some convenience methods, but I opted to give it a more consistent naming.

We could have both `Span` and `UnfinishedSpan` implement the same module `Datadog::SpanBase` or something like that for type check safety, although that feels like we're bending over backwards to support something we shouldn't.

**Comment by ericmustin** [comment]
So, i think to summarise is that i agree it's important to move upstream references out of `Span` so that it's basically just an exportable Struct which gets started/finished.   But have some questions about `SpanOperation`.

While we should definitely adopt a wrapper like `SpanOperation` (_in the branch it's referenced as `TraceOperation` and `SpanOperation`, is this meant to be two different things or the same?_), by moving to `SpanOperation`, I am a bit confused on where some of the additional details we've added to our current span `Context` will live.

In the branch I'm not sure whether [origin](https://github.com/DataDog/dd-trace-rb/blob/e391d2eb64d3c6151a4bdd2710c6a8c7c1d57457/lib/ddtrace/context.rb#L63), [sampling priority](https://github.com/DataDog/dd-trace-rb/blob/e391d2eb64d3c6151a4bdd2710c6a8c7c1d57457/lib/ddtrace/context.rb#L51), should live on the `SpanOperation`, remain on the `Context`, or some 3rd place like span tags?

One thing I'd like to understand, although I won't advocate a full blown endorsement of [w3c tracecontext](https://www.w3.org/TR/trace-context/) because I haven't really worked with it in production or anything, is how the components of `SpanOperation` outlined here map to that of `TraceContext` in w3c, specifically the `TraceFlags` and `TraceState` fields. I think some concepts such as `origin` map well to `TraceState` and `sampling_priority` map well to `TraceFlags`.

Anyways, just looking around at some other standards out there and trying to think about how ours map to them.

**Comment by delner** [comment]
@ericmustin `SpanOperation` is just a wrapper for the struct `Span`, with a `Context` and a `Tracer`, if available. It's not meant to be a fundamental rewrite of how context management works, rather a backwards-compatible stepping stone for our API,  such that we can make `Span` a bit simpler and enable context management changes later.

`SpanOperation` should be entirely interchangeable with the old usage of `Span`, assuming the use of duck typing.

As you've asked, the next step to take is to look more fundamentally at our context management naming and components. This will likely involve a sundering of `Context` and a change in terminology, but the details of this are yet undetermined.

**Comment by delner** [comment]
Opened #1675 which introduces `SpanOperation` changes

**Comment by marcotc** [comment]
This was done, and released in v1.0.0! 🎉
